controller/rancher/rancher.go

```diff
@@ -488,6 +550,238 @@ func (fetcher *RCertificateFetcher) UpdateEndpoints(lbSvc *metadata.Service, eps
+   Add a watcher for CertDir/DefaultCertDir also - upon events on these dirs clear the cache and enque.*/
+
+func (fetcher *RCertificateFetcher) ReadAllCertificatesFromDir(certDir string) ([]*config.Certificate, error) {
+	if _, err := os.Stat(certDir); os.IsNotExist(err) {
+		return nil, fmt.Errorf("Certificate Directory does not exist %v", certDir)
+	}
+	certs := []*config.Certificate{}
+
+	fetcher.mu.RLock()
+	currentCertD := fetcher.CertDir
+	fetcher.mu.RUnlock()
+
+	if currentCertD != "" && currentCertD == certDir && fetcher.CertsCache != nil && len(fetcher.CertsCache) > 0 {
+		//read from cache if certDir matches and certs are cached.
+		logrus.Debug("Reading certs from cache")
+	} else {
+		// init/re-init cache
+		fetcher.mu.Lock()
+		fetcher.removeWatches()
+
+		fetcher.CertsCache = make(map[string]*config.Certificate)
+		logrus.Debug("Reading certs from directory and starting watchers")
+		fetcher.TempCertsMap = make(map[string]*config.Certificate)
+		err := filepath.Walk(certDir, fetcher.readCertificate)
+		if err != nil {
+			fetcher.CertDir = ""
+			fetcher.CertsCache = nil
+			logrus.Errorf("Error %v reading certs from directory %v", err, certDir)
+		} else {
+			fetcher.CertDir = certDir
+			for path, cert := range fetcher.TempCertsMap {
+				fetcher.CertsCache[path] = cert
+			}
+			fetcher.addWatches()
+		}
+		fetcher.mu.Unlock()
+	}
+
+	fetcher.mu.RLock()
+	for _, value := range fetcher.CertsCache {
+		certs = append(certs, value)
+	}
+	fetcher.mu.RUnlock()
+
+	return certs, nil
+}
+
+func (fetcher *RCertificateFetcher) ReadDefaultCertificate(defaultCertDir string) (*config.Certificate, error) {
+	fetcher.mu.RLock()
+	currentDefCertD := fetcher.DefaultCertDir
+	fetcher.mu.RUnlock()
+
+	if currentDefCertD != defaultCertDir {
+		//defaultCertDir value changed, read it again.
+		fetcher.Watcher.Remove(defaultCertDir)
+		fetcher.mu.Lock()
+		fetcher.TempCertsMap = make(map[string]*config.Certificate)
+		err := filepath.Walk(defaultCertDir, fetcher.readCertificate)
+		if err != nil {
+			fetcher.DefaultCertDir = ""
+			fetcher.DefaultCert = nil
+			logrus.Errorf("Error %v reading defaultCertDir  %v", err, defaultCertDir)
+		} else {
+			fetcher.DefaultCertDir = defaultCertDir
+			for _, cert := range fetcher.TempCertsMap {
+				fetcher.DefaultCert = cert
+			}
+			fetcher.Watcher.Add(defaultCertDir)
+		}
+		fetcher.mu.Unlock()
+	}
+	return fetcher.DefaultCert, nil
+}
+
+func (fetcher *RCertificateFetcher) readCertificate(path string, f os.FileInfo, err error) error {
+	if f != nil && f.IsDir() {
+		logrus.Debugf("walking %v", path)
+		isCertFound := false
+		cert := config.Certificate{}
+		cert.Name = f.Name()
+		files, err := ioutil.ReadDir(path)
+		if err != nil {
+			return err
+		}
+		for _, file := range files {
+			if !file.IsDir() {
+				contentBytes, err := fetcher.evaluatueLinkAndReadFile(path, file.Name())
+				if err != nil {
+					return err
+				}
+				if file.Name() == "fullchain.pem" {
```

Comment:
we should probably declare fullchain/privatekey as constants somewhere. I believe I've seen them being used more than once in the code